### PR TITLE
MAP-1030 locations api create endpoint for use by change someones cell

### PR DESCRIPTION
**Pull request by Mjwillis**

### CODE REVIEW

File: src/main/resources/whereabouts/enabled.properties [hunk 1]
```diff
@@ -0,0 +1,117 @@
+# Whereabouts will be enabled for a prison if it's agencyId appears here, or in a properties file in whereabouts/patterns
```
[the remark this answers is not in the document]
Comment (GurnankCheema) [reply src/main/resources/whereabouts/enabled.properties]: yes, I did wonder that

File: src/main/kotlin/uk/gov/justice/digital/hmpps/locationsinsideprison/config/PropertiesConfiguration.kt [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+import org.springframework.context.annotation.Configuration
+import org.springframework.core.io.Resource
+
+@Configuration
+open class PropertiesConfiguration(
+  @Value("classpath:whereabouts/patterns/*.properties") private val resources: Array<Resource>,
+) {
+
+  @Bean
+  @Qualifier("whereaboutsGroups")
```
[the remark this answers is not in the document]
Comment (GurnankCheema) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/locationsinsideprison/config/PropertiesConfiguration.kt:16]: called it `locationPatternsProperties` to match the source file name

File: src/main/resources/whereabouts/patterns/BCI.properties [hunk 3]
```diff
@@ -0,0 +1,26 @@
+# Buckley Hall
```
[the remark this answers is not in the document]
Comment (GurnankCheema) [reply src/main/resources/whereabouts/patterns/BCI.properties:1]: changed dir to `locations/patterns`

**Review by Mjwillis — COMMENTED** [review]
This looks good, I would add tests now.  Suspect the values returned from the database will contain additional values we don't want.  we can look at that afterwards

File: src/test/kotlin/uk/gov/justice/digital/hmpps/locationsinsideprison/resource/LocationResourceIntTest.kt [hunk 4]
```diff
@@ -1312,7 +1351,102 @@ class LocationResourceIntTest : SqsIntegrationTestBase() {
...
+          .expectBody().json(
+            // language=json
+            """
+                [{
+                  "name":"Wing A",
+                  "key":"A",
+                  "children":[{"name":"Landing A","key":"A-1","children":[]}]
+                  },
+                  {
+                    "name":"Landing A",
```
Comment (GurnankCheema) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/locationsinsideprison/resource/LocationResourceIntTest.kt:1439]: This is unexpected. 
I expected the child to only be on line 1436 not duplicated at 1439 too

File: src/test/kotlin/uk/gov/justice/digital/hmpps/locationsinsideprison/resource/LocationResourceIntTest.kt [hunk 5]
```diff
@@ -212,11 +235,27 @@ class LocationResourceIntTest : SqsIntegrationTestBase() {
...
               {
                 "totalPages": 1,
-                "totalElements": 11,
+                "totalElements": 13,
                 "first": true,
                 "last": true,
-                "size": 11,
+                "size": 13,
                 "content": [
+                  {
```
Comment (GurnankCheema) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/locationsinsideprison/resource/LocationResourceIntTest.kt:243]: add NMI as it is not present in properties file

File: src/test/kotlin/uk/gov/justice/digital/hmpps/locationsinsideprison/resource/LocationResourceIntTest.kt [hunk 6]
```diff
@@ -72,6 +74,27 @@ class LocationResourceIntTest : SqsIntegrationTestBase() {
     prisonerSearchMockServer.resetAll()
     repository.deleteAll()
```
Comment (GurnankCheema) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/locationsinsideprison/resource/LocationResourceIntTest.kt:76]: add NMI as it is not included in .properties files

File: src/test/kotlin/uk/gov/justice/digital/hmpps/locationsinsideprison/resource/LocationResourceIntTest.kt [hunk 7]
```diff
@@ -1312,7 +1351,102 @@ class LocationResourceIntTest : SqsIntegrationTestBase() {
...
+                  "key":"A",
+                  "children":[{"name":"Landing A","key":"A-1","children":[]}]
+                  },
+                  {
+                    "name":"Landing A",
+                    "key":"A-1",
+                    "children":[]
+                  }
+                ]
+                        """,
```
Comment (GurnankCheema) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/locationsinsideprison/resource/LocationResourceIntTest.kt:1444]: Lines 1438 - 1442 are unexpected.
I expected the child to only be on line 1436 not duplicated at 1439 too

File: src/test/kotlin/uk/gov/justice/digital/hmpps/locationsinsideprison/resource/LocationResourceIntTest.kt (hunk 5 above)
Comment (GurnankCheema) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/locationsinsideprison/resource/LocationResourceIntTest.kt:243]: fix a breaking test

File: src/main/resources/whereabouts/patterns/BCI.properties [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+BCI_A-Wing=BCI-A-.+
+BCI_A-Wing_Landing\ 1+2=BCI-A-[1-2]-.+
+BCI_A-Wing_Landing\ 3+4=BCI-A-[3-4]-.+
+BCI_A-Wing_Landing\ 1=BCI-A-1-.+
+BCI_A-Wing_Landing\ 2=BCI-A-2-.+
+BCI_A-Wing_Landing\ 3=BCI-A-3-.+
+BCI_A-Wing_Landing\ 4=BCI-A-4-.+
+BCI_B-Wing=BCI-B-.+
+BCI_B-Wing_Landing\ 1+2=BCI-B-[1-2]-.+
```
Comment (marcusaleman) [inline src/main/resources/whereabouts/patterns/BCI.properties]: @Mjwillis @GurnankCheema 
I had a quick HUDDLE with @GurnankCheema  in order to understand the properties why are so long and in my opinion repetitive.
as per my understanding is cause NOMIS is using those patterns and it is a work around it.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/locationsinsideprison/resource/LocationResourceIntTest.kt (hunk 7 above)
Comment (Mjwillis) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/locationsinsideprison/resource/LocationResourceIntTest.kt:1444]: You can loading all locations not just the wings hence why

Comment (Mjwillis) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/locationsinsideprison/resource/LocationResourceIntTest.kt:1444]: I've corrected

**Comment by Mjwillis** [comment]
I've fixed the test.  I'll merge the changes for now and you can pick up next week @GurnankCheema

**Pull request merged**